pydm/widgets/rules.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import functools
+
+from ..PyQt.QtCore import QThread, pyqtSignal
+from ..PyQt.QtGui import QApplication
+
+from .channel import PyDMChannel
+from ..utilities import is_pydm_app
+
+import numpy as np
+from math import *
+
+logger = logging.getLogger(__name__)
+
+
+class RulesEngine(QThread):
+    """
+    RulesEngine inherits from QThread and is responsible for monitoring the
+    channels associated with a rule, evaluate the expression when new values
+    arrive and emit the signal for the widget so it can be updated properly.
+
+    Parameters
+    ----------
+    rule_map : dict
+        The dictionary containing the rule information needed for the engine.
+
+    Signals
+    -------
+    rule_signal : dict
+        Emitted when a new value for the property is calculated by the engine.
+    """
+    rule_signal = pyqtSignal(dict)
+
+    def __init__(self, rule_map):
+        QThread.__init__(self)
+        if rule_map is None or not isinstance(rule_map, dict):
+            raise ValueError("Invalid format for rule_map. Dictionary expected")
+        # Enables termination of the current thread
+        self.setTerminationEnabled(True)
+
+        # Reference to App so we can establish the connection with the channel
+        self.app = QApplication.instance()
+
+        # Flag to control whether or not we should evaluate the expression
+        self.should_calculate = False
+
+        # Definitions for this Action Thread
+        self.rule_map = rule_map
+        self.name = rule_map.get('name', None)
+        self.property = rule_map.get('property', None)
+        self.expression = rule_map.get('expression', None)
+        self.channels_list = rule_map.get('channels', [])
+
+        # Buffer for data coming from the channels
+        self.channels_connection = [False] * len(self.channels_list)
+        self.channels_value = [None] * len(self.channels_list)
+
+        self.channels = []
+
+        for idx, ch in enumerate(self.channels_list):
+            partial_connection = functools.partial(self.channel_conn_callback,
+                                                   idx, ch['channel'],
+                                                   ch['trigger'])
+            partial_value = functools.partial(self.channel_value_callback,
+                                              idx, ch['channel'], ch['trigger'])
+            c = PyDMChannel(ch['channel'],
+                            connection_slot=partial_connection,
+                            value_slot=partial_value)
+            if is_pydm_app():
+                self.app.add_connection(c)
+            self.channels.append(c)
```

Comment:
Might be nice if these channels were also available from the parent widget. Really if you have a connection and a rule you should see two channels when you access `widget.channels()`

Reply:
Got it.... I was imagining to separate the Rules from the widget... they are almost stand alone with the exception of the property `rules` and the signal...
@mattgibbs what do you think?